base/compiler/optimize.jl

```diff
@@ -453,9 +457,262 @@ function finish(interp::AbstractInterpreter, opt::OptimizationState,
+                    all_retpaths_consistent = false
+                    # Check if there are potential throws that require
+                elseif conditional_successors_may_throw(lazypostdomtree, ir, bb)
+                    all_retpaths_consistent = false
+                else
+                    (cfg, domtree) = get_augmented_domtree()
+                    for succ in iterated_dominance_frontier(cfg, BlockLiveness((sa, sb), nothing), domtree)
+                        if succ == length(cfg.blocks)
+                            # Phi node in the virtual exit -> We have a conditional
+                            # return. TODO: Check if all the retvals are egal.
+                            all_retpaths_consistent = false
+                        else
+                            visit_bb_phis!(ir, succ) do phiidx::Int, phi::PhiNode
+                                push!(inconsistent, phiidx)
+                            end
+                        end
+                    end
+                end
+            end
+        end
+
+        return true
+    end
+    if !scan!(scan_stmt!, scanner, true)
+        if !all_retpaths_consistent
+            # No longer any dataflow concerns, just scan the flags
+            scan!(scanner, false) do inst, idx, lstmt, bb
+                scan_non_dataflow_flags!(inst)
+                return true
+            end
+        else
+            scan!(scan_stmt!, scanner, true)
+            complete!(tpdum); push!(scanner.bb_ip, 1)
+            populate_def_use_map!(tpdum, scanner)
+            for def in inconsistent
+                for use in tpdum[def]
+                    if !(use in inconsistent)
+                        push!(inconsistent, use)
+                        append!(stmt_ip, tpdum[use])
+                    end
+                end
+            end
+
+            stmt_ip = BitSetBoundedMinPrioritySet(length(ir.stmts))
+            while !isempty(stmt_ip)
+                idx = popfirst!(stmt_ip)
+                inst = ir[SSAValue(idx)]
+                stmt = inst[:inst]
+                if is_getfield_boundscheck(inst)
+                    any_non_boundscheck_inconsistent = false
+                    for i = 1:(length(stmt.args)-1)
+                        val = stmt.args[i]
+                        if isa(val, SSAValue)
+                            any_non_boundscheck_inconsistent |= val.id in inconsistent
+                            any_non_boundscheck_inconsistent && break
+                        end
+                    end
+                    any_non_boundscheck_inconsistent || continue
+                elseif isa(stmt, ReturnNode)
+                    all_retpaths_consistent = false
+                else isa(stmt, GotoIfNot)
+                    bb = block_for_inst(ir, idx)
+                    sa, sb = ir.cfg.blocks[bb].succs
+                    for succ in iterated_dominance_frontier(ir.cfg, BlockLiveness((sa, sb), nothing), get!(lazydomtree))
```

Comment:
```suggestion
                    for succ in iterated_dominance_frontier(ir.cfg, BlockLiveness(ir.cfg.blocks[bb].succs, nothing), get!(lazydomtree))
```
and then we can keep typing `BlockLiveness.def_bbs::Vector{Int}`.